OrbitCore/LinuxPerfEventProcessor.cpp

```diff
@@ -12,47 +12,89 @@
 
 #include "Profiling.h"
 
-void LinuxPerfEventProcessor::Push(std::unique_ptr<LinuxPerfEvent> event) {
+void PerfEventQueue::PushEvent(int origin_fd,
+                               std::unique_ptr<LinuxPerfEvent> event) {
+  if (fd_event_queues_.count(origin_fd) > 0) {
+    std::shared_ptr<std::queue<std::unique_ptr<LinuxPerfEvent>>> event_queue =
+        fd_event_queues_.at(origin_fd);
+    assert(!event_queue->empty());
+    // Fundamental assumption: events from the same file descriptor come already
```

Comment:
Is this really the case?  If I understand correctly, all events from an instrumented function are piped to the same file descriptor.  This function could be called concurrently from multiple threads.  Couldn't it be the case that thread A gets preempted just before writing to the event buffer, then thread B generates and writes new events before thread A resumes and writes its older events?

Reply:
My understanding is that, as the kernel is taking care of the data collection, the writes to the ring buffer should be atomic and the situation you describe shouldn't happen. But indeed I wouldn't know how to prove it without extensive testing.

Reply:
What I'm describing is not a a problem with the atomicity of the writes, but rather that a thread could be kicked out of a core right before the write happens.  Another thread could then write new events to the file before the initial thread is re-scheduled and writes the now out-of-order data.

This code being so crucial I think we must write stress tests to first confirm that out-of-order events could happen and then properly handle the situation.

Reply:
My point was that since the events are not timestamped or written by the threads but by the kernel itself, I would expect that the kernel don't de-schedule itself right after having registered the timestamp but before writing to the ring buffer but perform the two operations atomically.

Reply:
So what are we doing about this? This improvement is not critical or a priority so I'm not going to write stress tests now.
We can: close it for now; merge it and revert it if needed; keep both implementations around so that we can easily switch between the two.

Reply:
Hi Davide, having both implementations would be great.  I was planning on writing some stress tests anyway so we could use them to compare both solutions.  I'd really like to know if the case I was mentioning can actually occur.  If it does, we should think about how to recover.

Reply:
So now we have both, LinuxPerfEventProcessor (the original one) and LinuxPerfEventProcessor2 (the name can be reworked...). It should be possible to switch between the two just by changing the type of uprobe_event_processor in LinuxEventTracer.cpp.

Reply:
Ok great, thanks for the change.